compiler/rustc_monomorphize/src/partitioning.rs

```diff
@@ -1102,7 +1111,66 @@ where
     }
 }
 
-fn collect_and_partition_mono_items(tcx: TyCtxt<'_>, (): ()) -> (&DefIdSet, &[CodegenUnit<'_>]) {
+pub(crate) fn adjust_activity_to_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    fn_ty: Ty<'tcx>,
+    da: &mut Vec<DiffActivity>) {
+    if !fn_ty.is_fn() {
+        // Error?
+        return;
+    }
+    let fnc_binder: ty::Binder<'_, ty::FnSig<'_>> = fn_ty.fn_sig(tcx);
+
+    // If rustc compiles the unmodified primal, we know that this copy of the function
+    // also has correct lifetimes. We know that Enzyme won't free the shadow too early
+    // (or actually at all), so let's strip lifetimes when computing the layout.
+    // Recommended by compiler-errors:
+    // https://discord.com/channels/273534239310479360/957720175619215380/1223454360676208751
+    let x = tcx.instantiate_bound_regions_with_erased(fnc_binder);
+    let mut new_activities = vec![];
+    let mut new_positions = vec![];
+    for (i, ty) in x.inputs().iter().enumerate() {
+        if ty.is_unsafe_ptr() || ty.is_ref() || ty.is_box() {
+            if ty.is_fn_ptr() {
+                unimplemented!("what to do whith fn ptr?");
+            }
```

Comment:
We're in a limbo, because I first implemented type/safety checks in the autodiff macro, then I realized I want more type checks than a rustc_builtin_macro can offer, but didn't want to hold back the upstream on yet another rewrite, after just having finished another one going from proc-macro's to builtin-macros.

This function (in it's original implementation, not this striped-down one), was traversing all types through indirections recursively, something with the autodiff_macro never will be able to do. I might want to support fn pointers one day, but I guess for now I should just add a propper error message about this, since it could come from an honest user mistake, that I just didn't catch earlier.

Reply:
A proper error message for this case is fine, yeah.